core/commands/dag/dag.go

```diff
@@ -241,3 +261,248 @@ var DagResolveCmd = &cmds.Command{
 	},
 	Type: ResolveOutput{},
 }
+
+type importResult struct {
+	roots map[cid.Cid]bool
+	err   error
+}
+
+var DagImportCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Import the contents of .car files",
+		ShortDescription: `
+'ipfs dag import' parses .car files and adds all objects to the blockstore.
+Currently only car files with version 1 are supported, without particular
+hard limits on the contents (so called car-determinism is not enforced).
+
+By default, after all supplied (one or more) car files have been processed,
+an attempt is made to pin recursively each individual root specified in any car
+headers. This is done at the end of the import, allowing pinning to take place
+over the contents of all cars (a dag scattered over multiple files) plus
+anything present in the blockstore.
+
+During pinning no network requests are made, with no option to change that.
+`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.FileArg("path", true, true, "The path of a .car file.").EnableStdin(),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption(silentOptionName, "No output."),
+		cmds.BoolOption(pinRootsOptionName, "Pin optional roots listed in the .car headers after importing.").WithDefault(true),
+	},
+	Type: CarImportOutput{},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+
+		node, err := cmdenv.GetNode(env)
+		if err != nil {
+			return err
+		}
+
+		api, err := cmdenv.GetApi(env, req)
+		if err != nil {
+			return err
+		}
+
+		// on import ensure we do not reach out to the network for any reason
+		// if a pin based on what is imported + what is in the lockstore
+		// isn't possible: tough luck
+		api, _ = api.WithOptions(options.Api.Offline(true))
+
+		// grab a gc lock so that regardless of the size of the streamed-in cars
+		// nothing will disappear on us before we had a chance to pin the roots
+		// at the very end
+		// This is especially important for use cases like dagger:
+		//    ipfs dag import $( ... | ipfs-dagger --stdout=carfifos )
+		//
+		unlocker := node.Blockstore.PinLock()
+		defer func() { unlocker.Unlock() }()
+
+		doPinRoots, _ := req.Options[pinRootsOptionName].(bool)
+
+		retCh := make(chan importResult, 1)
+		go importWorker(req, res, api, retCh)
```

Comment:
you don't get anything from the go routine complexity in this construction. is there another use or reason for this?

Reply:
Another PR will include progress feedback utilizing this.

Reply:
Simply emitting a response every time we add a block is probably going to be simpler than having a separate goroutine.

Reply:
Wouldn't it be **way** too expensive to JSON encode a counter message on every block for multi-thousand block cars. This is why the original implementation used a timer-ticker based emitter. If you feel strongly about it: I will switch to emitting on every block, but ewwww.

Reply:
Knowing all the CIDs in a CAR would be generally useful. But I see your point. We'd only want to enable that with a flag in special cases.

Reply:
A ticking progress reporter goroutine also has the nice attribute of being in a position to notice (and report) if some operations stall.